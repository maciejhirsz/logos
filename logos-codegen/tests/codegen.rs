use std::{error::Error, io, path::PathBuf};

#[rstest::rstest]
#[case("simple")]
#[case("no_error_lut")]
pub fn test_codegen(#[case] fixture: &str) -> Result<(), Box<dyn Error>> {
    let mut fixture_dir = PathBuf::new();
    fixture_dir.push(env!("CARGO_MANIFEST_DIR"));
    fixture_dir.push("tests");
    fixture_dir.push("data");
    fixture_dir.push(fixture);

    let input = fixture_dir.join("input.rs");
    fixture_dir.push("output.rs");
    let output_file = fixture_dir;

    let input = std::fs::read_to_string(input)?;
    let output = std::fs::read_to_string(&output_file)?;

    let generated = logos_codegen::generate(input.parse()?);
    let generated = generated.to_string();

    if std::env::var("BLESS_CODEGEN").is_ok_and(|value| value == "1") {
        std::fs::write(&output_file, &generated)?;
        return Ok(());
    }

    assert_eq!(generated, output, "Codegen test failed: `{fixture}`, run tests again with env var `BLESS_CODEGEN=1` to bless these changes");

    Ok(())
}
