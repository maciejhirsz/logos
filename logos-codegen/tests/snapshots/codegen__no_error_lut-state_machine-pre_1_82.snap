---
source: logos-codegen/tests/codegen.rs
expression: formatted
---
impl<'s> ::logos::Logos<'s> for Token {
    type Error = ();
    type Extras = ();
    type Source = [u8];
    fn lex(
        lex: &mut ::logos::Lexer<'s, Self>,
    ) -> std::option::Option<std::result::Result<Self, Self::Error>> {
        use ::logos::internal::{
            LexerInternal, CallbackRetVal, CallbackResult, SkipRetVal, SkipResult,
        };
        use std::result::Result as _Result;
        use std::option::Option as _Option;
        use ::logos::Logos;
        type _Lexer<'s> = ::logos::Lexer<'s, Token>;
        macro_rules! _fast_loop {
            ($lex:ident, $test:ident, $offset:ident) => {
                'fast_loop : { while let Some(arr) = $lex .read:: < & [u8; 8usize] >
                ($offset) { if $test (arr[0usize]) { $offset += 0usize; break 'fast_loop;
                } if $test (arr[1usize]) { $offset += 1usize; break 'fast_loop; } if
                $test (arr[2usize]) { $offset += 2usize; break 'fast_loop; } if $test
                (arr[3usize]) { $offset += 3usize; break 'fast_loop; } if $test
                (arr[4usize]) { $offset += 4usize; break 'fast_loop; } if $test
                (arr[5usize]) { $offset += 5usize; break 'fast_loop; } if $test
                (arr[6usize]) { $offset += 6usize; break 'fast_loop; } if $test
                (arr[7usize]) { $offset += 7usize; break 'fast_loop; } $offset += 8usize;
                } while let Some(byte) = $lex .read:: < u8 > ($offset) { if $test (byte)
                { break 'fast_loop; } $offset += 1; } }
            };
        }
        #[inline]
        fn _make_error<'s>(lex: &mut _Lexer<'s>) -> <Token as Logos<'s>>::Error {
            <Token as Logos<'s>>::Error::default()
        }
        #[derive(Clone, Copy)]
        enum LogosState {
            state0,
            state1,
            state10,
            state11,
            state12,
            state13,
            state14,
            state15,
            state16,
            state17,
            state18,
            state19,
            state2,
            state3,
            state4,
            state5,
            state6,
            state7,
            state8,
            state9,
        }
        let mut state = LogosState::state0;
        let mut offset = lex.offset();
        loop {
            match state {
                LogosState::state0 => {
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {
                        const TABLE: [_Option<LogosState>; 256] = [
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state2),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state1),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state4),
                            Some(LogosState::state5),
                            Some(LogosState::state6),
                            Some(LogosState::state6),
                            Some(LogosState::state6),
                            Some(LogosState::state6),
                            Some(LogosState::state6),
                            Some(LogosState::state6),
                            Some(LogosState::state6),
                            Some(LogosState::state6),
                            Some(LogosState::state6),
                            Some(LogosState::state6),
                            Some(LogosState::state6),
                            Some(LogosState::state6),
                            Some(LogosState::state7),
                            Some(LogosState::state6),
                            Some(LogosState::state6),
                            Some(LogosState::state8),
                            Some(LogosState::state9),
                            Some(LogosState::state9),
                            Some(LogosState::state9),
                            Some(LogosState::state10),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                            Some(LogosState::state3),
                        ];
                        let next_state = TABLE[byte as usize];
                        if let Some(next_state) = next_state {
                            offset += 1;
                            state = next_state;
                            continue;
                        }
                    } else {
                        if lex.offset() == offset {
                            return None;
                        }
                    }
                    lex.end_to_boundary(offset + if other.is_some() { 1 } else { 0 });
                    return Some(Err(_make_error(lex)));
                }
                LogosState::state1 => {
                    lex.end(offset);
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {} else {}
                    return Some(Ok(Token::AnyUnicode));
                }
                LogosState::state2 => {
                    lex.end(offset);
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {} else {}
                    return Some(Ok(Token::Newline));
                }
                LogosState::state3 => {
                    lex.end(offset);
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {} else {}
                    return Some(Ok(Token::Any));
                }
                LogosState::state4 => {
                    lex.end(offset);
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {
                        if (matches!(byte, 128u8..= 191u8)) {
                            offset += 1;
                            state = LogosState::state1;
                            continue;
                        }
                    } else {}
                    return Some(Ok(Token::Any));
                }
                LogosState::state5 => {
                    lex.end(offset);
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {
                        if (matches!(byte, 160u8..= 191u8)) {
                            offset += 1;
                            state = LogosState::state13;
                            continue;
                        }
                    } else {}
                    return Some(Ok(Token::Any));
                }
                LogosState::state6 => {
                    lex.end(offset);
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {
                        if (matches!(byte, 128u8..= 191u8)) {
                            offset += 1;
                            state = LogosState::state13;
                            continue;
                        }
                    } else {}
                    return Some(Ok(Token::Any));
                }
                LogosState::state7 => {
                    lex.end(offset);
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {
                        if (matches!(byte, 128u8..= 159u8)) {
                            offset += 1;
                            state = LogosState::state13;
                            continue;
                        }
                    } else {}
                    return Some(Ok(Token::Any));
                }
                LogosState::state8 => {
                    lex.end(offset);
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {
                        if (matches!(byte, 144u8..= 191u8)) {
                            offset += 1;
                            state = LogosState::state12;
                            continue;
                        }
                    } else {}
                    return Some(Ok(Token::Any));
                }
                LogosState::state9 => {
                    lex.end(offset);
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {
                        if (matches!(byte, 128u8..= 191u8)) {
                            offset += 1;
                            state = LogosState::state12;
                            continue;
                        }
                    } else {}
                    return Some(Ok(Token::Any));
                }
                LogosState::state10 => {
                    lex.end(offset);
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {
                        if (matches!(byte, 128u8..= 143u8)) {
                            offset += 1;
                            state = LogosState::state12;
                            continue;
                        }
                    } else {}
                    return Some(Ok(Token::Any));
                }
                LogosState::state11 => {
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {} else {}
                    return Some(Ok(Token::Any));
                }
                LogosState::state12 => {
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {
                        if (matches!(byte, 128u8..= 191u8)) {
                            offset += 1;
                            state = LogosState::state13;
                            continue;
                        }
                    } else {}
                    return Some(Ok(Token::Any));
                }
                LogosState::state13 => {
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {
                        if (matches!(byte, 128u8..= 191u8)) {
                            offset += 1;
                            state = LogosState::state1;
                            continue;
                        }
                    } else {}
                    return Some(Ok(Token::Any));
                }
                LogosState::state14 => {
                    lex.end(offset);
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {} else {}
                    return Some(Ok(Token::AnyUnicode));
                }
                LogosState::state15 => {
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {} else {}
                    return Some(Ok(Token::AnyUnicode));
                }
                LogosState::state16 => {
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {
                        if (matches!(byte, 128u8..= 191u8)) {
                            offset += 1;
                            state = LogosState::state1;
                            continue;
                        }
                    } else {}
                    return Some(Ok(Token::Any));
                }
                LogosState::state17 => {
                    lex.end(offset);
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {} else {}
                    return Some(Ok(Token::AnyUnicode));
                }
                LogosState::state18 => {
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {} else {}
                    return Some(Ok(Token::Newline));
                }
                LogosState::state19 => {
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {} else {}
                    return Some(Ok(Token::AnyUnicode));
                }
            }
        }
    }
}
